perfdash/perfdash.go

```diff
@@ -134,37 +140,94 @@ func parseTestOutput(scanner *bufio.Scanner, buildNumber int, resources sets.Str
 	return hist
 }
 
-func getLatencyData(client *jenkins.JenkinsClient, job string) (TestToBuildData, sets.String, sets.String, error) {
+// returns build number of the next run greater than *startFrom
+// TODO: if anyone has any idea how to do it better, please do
+func binSearch() int {
+	start := *startFrom
+	end := *startFrom + 1
+	// search from some build number without result
+	response, err := http.Get(fmt.Sprintf("%v%v%v", urlPrefix, start, urlSuffix))
+	if err != nil {
+		panic(err)
+	}
+	if response.StatusCode != 200 {
+		return start
+	}
+	response, err = http.Get(fmt.Sprintf("%v%v%v", urlPrefix, end, urlSuffix))
+	if err != nil {
+		panic(err)
+	}
+	if response.StatusCode != 200 {
+		return end
+	}
+	inc := 1
+	for response.StatusCode == 200 {
+		inc *= 2
+		start = end
+		end += inc
+		response, err = http.Get(fmt.Sprintf("%v%v%v", urlPrefix, end, urlSuffix))
+		if err != nil {
+			panic(err)
+		}
+	}
+	// We know that data exists for start and does not exist for end. Now binary search.
+	for start+1 < end {
+		candidate := (start + end) / 2
+		response, err = http.Get(fmt.Sprintf("%v%v%v", urlPrefix, end, urlSuffix))
+		if err != nil {
+			panic(err)
+		}
+		if response.StatusCode == 200 {
+			start = candidate + 1
+		} else {
+			end = candidate
+		}
+	}
+	// We're left with at most two candidates
+	if start == end {
+		return start
+	}
+	response, err = http.Get(fmt.Sprintf("%v%v%v", urlPrefix, start, urlSuffix))
+	if err != nil {
+		panic(err)
+	}
+	if response.StatusCode == 200 {
+		return end
+	}
+	return start
+}
+
+func getLatencyDataFormGCS() (TestToBuildData, sets.String, sets.String, error) {
 	buildLatency := TestToBuildData{}
 	resources := sets.NewString()
 	methods := sets.NewString()
 
-	queue, err := client.GetJob(job)
-	if err != nil {
-		return buildLatency, resources, methods, err
+	buildNumber := *startFrom
+	lastBuildNo := binSearch()
+	if buildNumber < lastBuildNo-100 {
+		buildNumber = lastBuildNo - 100
 	}
 
-	for ix := range queue.Builds {
-		build := queue.Builds[ix]
-		if build.Number < *startFrom {
-			continue
-		}
-		reader, err := client.GetConsoleLog(job, build.Number)
+	for ; buildNumber < lastBuildNo; buildNumber++ {
+		testDataResponse, err := http.Get(fmt.Sprintf("%v%v%v", urlPrefix, buildNumber, urlSuffix))
 		if err != nil {
-			fmt.Printf("error getting logs: %v\n", err)
-			continue
+			panic(err)
```

Comment:
I think that doing panic() is risky. I think you should log error and continue here